lib/logdna/client.rb

```diff
@@ -49,27 +51,23 @@ def schedule_flush
         sleep(@exception_flag ? @retry_timeout : @flush_interval)
         flush if @flush_scheduled
       }
-      Thread.new { start_timer.call }
+
+      Concurrent::Future.execute({ executor: @thread_pool }) { start_timer.call }
     end
 
     def write_to_buffer(msg, opts)
-      if @lock.try_lock
+      @lock.synchronize do
         processed_message = process_message(msg, opts)
         new_message_size = processed_message.to_s.bytesize
         @buffer.push(processed_message)
         @buffer_byte_size += new_message_size
-        @flush_scheduled = true
-        @lock.unlock
 
         if @flush_limit <= @buffer_byte_size
-          flush
+          Concurrent::Future.execute({ executor: @thread_pool }) { flush }
         else
+          @flush_scheduled = true
           schedule_flush
         end
-      else
-        @side_message_lock.synchronize do
-          @side_messages.push(process_message(msg, opts))
-        end
```

Comment:
without this, `side_messages` appears to be only used for pushing messages from failed requests, is that correct?
with this change `write_to_buffer` will just block until a lock is available rather that pushing new messages into the side buffer.

Reply:
Yeah... and I think we need to re-work how the buffer & side messages work.  The way they're currently handled, the threadpool will still access `@buffer` and `@side_messages`, which isn't thread-safe.  The locking will probably also cause unintended consequences.

I'm going to change this logic.